src/interpreter/Func.kt

```diff
@@ -0,0 +1,73 @@
+package interpreter
+
+import core.*
+
+fun prettyPrint(arg : Value) : String = when(arg) {
+    is VInt -> arg.value.toString()
+    is VString -> arg.value
+    is VBool -> arg.value.toString()
+    is VArray -> arg.value.joinToString(prefix = "[", postfix = "]", transform = ::prettyPrint)
+    is VUnit -> "unit"
+}
+
+// Function objects
+sealed class Func(val id : String, val args : List<AnnotatedVar>, val returnType : Type) {
+    override fun equals(other : Any?) =
+            other is Func && id == other.id && args == other.args && returnType == returnType
+
+    override fun hashCode() = 31 * (31 * id.hashCode() + args.hashCode()) + returnType.hashCode()
+    operator fun component1() = id
+    operator fun component2() = args
+    operator fun component3() = returnType
+}
+
+class TommyFunc(id : String, args : List<AnnotatedVar>, returnType : Type, val statements : Body,
+                val environment : Map<String, Value>)
```

Comment:
I don't see why TommyFunc should have its own environment as these objects are only created at parse time.
Also, why is this data class expanded all the way out, seems unnecessary.

Reply:
TommyFunc is created at evaluation time, specifically because it needs that environment. Data classes can only have val parameters to their conductor, but I wanted Func to have val parameters and then TommyFunc to just use those. I don't really care about that though, you can remove the body of Func and make TommyFunc a data class

Reply:
oh I see now why the data class is expanded.